### PR TITLE
Feature: Make a book catalog with new design JC-597

### PR DESCRIPTION
**Pull request by manpham-saritasa**
Make a book catalog page based on new design with HTML and CSS [JC-597](https://saritasa.atlassian.net/browse/JC-597)

NOTE: 
The reason why this PR only has one commit is because some of the previous commits were committed by my personal GitHub account. In order to fix the problem, I use git reset --soft and stash all of the commits into one commit to ensure that it is committed by the correct account.

[JC-597]: https://saritasa.atlassian.net/browse/JC-597?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/vanilla/src/pages/index.html [hunk 1]
```diff
@@ -6,11 +6,149 @@
...
+                <img src="./assets/images/BookCover/blue-period.png" alt="Blue Period" />
+              </div>
+              <div class="book-card__content">
+                <span class="book-card__name font-body-small">Blue Period</span>
+              </div>
+            </article>
+          </div>
+        </section>
+
+        <section class="book-group">
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/pages/index.html:50]: In the lecture for russian team, I asked the guys to use grid for cards view. I understand that you didn't attend to lecture, so I'm asking you now.

The problem with flexbox, is that it's imposible to implement cards wrapping without empty space on the side (second image's width is slightly smaller): 
image
Screenshot 2024-07-04 at 12 57 00

You can use this article:
https://ishadeed.com/article/css-grid-minmax/

TLDR: use `grid-template-columns: repeat(auto-fill, minmax(180px, 1fr));`

File: apps/vanilla/src/global/components.css [hunk 2]
```diff
@@ -1,28 +1,59 @@
...
+  }
 
-  margin: 80px auto 0;
-  max-width: var(--default-layout-width);
+  .footer__information {
+    gap: 0;
+    width: 100%;
+    justify-content: space-between;
+  }
 }
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/global/components.css:59]: It's good practice to place style blocks in media queries close to regular style blocks.

I mean this:
```css
.footer {...}
.footer__information {...}
@media only screen and (max-width: 767px) {
  .footer {...}
  .footer__information {...}
}
```

This way it's easier to grasp context and developer don't need to scroll back and forth.

File: apps/vanilla/src/global/base.css [hunk 3]
```diff
@@ -1,4 +1,11 @@
 html {
   font-size: 20px;
   font-family: system-ui, sans-serif;
+  box-sizing: border-box;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/global/base.css:4]: It's okay to use it now. But let's not change `box-sizing` for all elements in future tasks for Angular and React, because we don't usually do this in our projects.

File: apps/vanilla/src/global/components.css [hunk 4]
```diff
@@ -1,28 +1,59 @@
...
+}
+
+/* Checkbox input */
+input[type="checkbox"] {
+  accent-color: var(--primary-color);
+}
+
+/* Container */
+.container {
+  max-width: 1280px;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/global/components.css:34]: I think it's good idea to extract it into variable, like you did for header and footer.

File: apps/vanilla/src/pages/index.css [hunk 5]
```diff
@@ -1,5 +1,208 @@
...
+.book-card__cover {
+  max-width: 100%;
+  aspect-ratio: 2/3;
+  border-radius: var(--border-radius-lg);
+  border: 1px solid var(--border-color);
+  overflow: hidden;
+}
+
+.book-card__cover img {
+  object-fit: fill;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/pages/index.css:57]: `fill` is distorting/stretching image, let's use `cover`.

```suggestion
  object-fit: cover;
```

File: apps/vanilla/src/pages/index.css [hunk 6]
```diff
@@ -1,5 +1,208 @@
...
+/* Filter */
+.sidebar {
+  --sidebar-width: 320px;
+
+  position: relative;
+  width: var(--sidebar-width);
+}
+
+.sidebar__filter-wrapper {
+  position: absolute;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/pages/index.css:86]: Why do you use it here? `position: absolute` should be used as less as possible, and I don't any reason to use in this case.

File: apps/vanilla/src/pages/index.css [hunk 7]
```diff
@@ -1,5 +1,208 @@
...
+}
+
+.filter__form {
+  display: flex;
+  flex-direction: column;
+  gap: var(--spacing-lg);
+  margin-top: var(--spacing-lg);
+}
+
+.filter__form .search-bar {
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/pages/index.css:101]: BEM conceptually dissallows nested classes as selectors.

Your case should be rewritten this way to align with BEM methodology.
```css
.filter__seach {}
```

File: apps/vanilla/src/pages/index.css [hunk 8]
```diff
@@ -1,5 +1,208 @@
...
+  border-radius: var(--border-radius-md);
+  max-width: 100%;
+  background-color: var(--surface-overlay-color);
+}
+
+.filter__select select {
+  border: none;
+  background-color: transparent;
+  width: 100%;
+}
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/pages/index.css:129]: I think select and input should be **Block** (from BEM), and should be styled globally.

File: apps/vanilla/src/pages/index.css (hunk 7 above)
Comment (artem-sotnichenko-saritasa) [reply apps/vanilla/src/pages/index.css:101]: Basically, you always should use only 1 class by BEM methodology.

The only exception is state related styles, example:
```css
.partent:hover .parent__child {...}
```

File: apps/vanilla/src/pages/index.html (hunk 1 above)
Comment (artem-sotnichenko-saritasa) [reply apps/vanilla/src/pages/index.html:50]: It also would work good on mobile screens

**Review by artem-sotnichenko-saritasa — CHANGES_REQUESTED** [review]

**Comment by artem-sotnichenko-saritasa** [comment]
Also, please update styles for select and checkbox to match designs.

File: apps/vanilla/src/pages/index.css (hunk 6 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:86]: Because I want the filter block to be fixed but still related to the parent which is containing it. Hence the mentioned implementation.

Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:86]: My idea is to make the .filter to be "position:fixed" but still related to the parent element. Hence the mentioned implementation.

Here is the documentation: https://stackoverflow.com/a/11833892

File: apps/vanilla/src/pages/index.css (hunk 8 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:129]: Do you mean <input type="checkbox" /> element only ?

Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:129]: Do you mean the inputs with the type="checkbox" only?

File: apps/vanilla/src/pages/index.html (hunk 1 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.html:50]: Thanks! The problem has been resolved

File: apps/vanilla/src/global/components.css (hunk 2 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/global/components.css:59]: Thanks! The problem has been resolved

File: apps/vanilla/src/global/components.css (hunk 4 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/global/components.css:34]: Thanks! The problem has been resolved

File: apps/vanilla/src/pages/index.css (hunk 5 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:57]: Thanks! The problem has been resolved

File: apps/vanilla/src/pages/index.css (hunk 7 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:101]: Thanks! The problem has been resolved

File: apps/vanilla/src/global/base.css (hunk 3 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/global/base.css:4]: So should I remove it or change its value?

Comment (artem-sotnichenko-saritasa) [reply apps/vanilla/src/global/base.css:4]: Keep it in vanilla.

File: apps/vanilla/src/pages/index.css (hunk 6 above)
Comment (artem-sotnichenko-saritasa) [reply apps/vanilla/src/pages/index.css:88]: Hmm, got it, it just don't needed in this particular case. You can just remove it, and `position fixed;` will still be working.

File: apps/vanilla/src/global/base.css (hunk 3 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/global/base.css:4]: Thank you!

File: apps/vanilla/src/pages/index.css (hunk 8 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css]: Thank you! The problem has been resolved

File: apps/vanilla/src/pages/index.css (hunk 6 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/pages/index.css:88]: Thanks a lot! The problem has been resolved

File: apps/vanilla/src/global/variables.css [hunk 9]
```diff
@@ -1,5 +1,67 @@
...
+.font-body-small {
+  font-size: var(--font-size-md);
+  line-height: var(--line-height-md);
+  font-weight: var(--font-weight-medium);
+}
+
+.font-logo {
+  font-size: var(--font-size-xl);
+  line-height: var(--line-height-md);
+  font-weight: var(--font-weight-bold);
```
Comment (maxineo) [inline apps/vanilla/src/global/variables.css:66]: These are not variables so lets move them to a separate file.

**Review by maxineo — APPROVED** [review]

File: apps/vanilla/src/global/variables.css (hunk 9 above)
Comment (manpham-saritasa) [reply apps/vanilla/src/global/variables.css:66]: How about I move them to another file called "styles.css" ?

Comment (maxineo) [reply apps/vanilla/src/global/variables.css:66]: I think `typography.css` will be fine.

Comment (manpham-saritasa) [reply apps/vanilla/src/global/variables.css]: TYSM! The problem has been resolved

**Review by EgorRyazanov — APPROVED** [review]
Good job!

**Review by maxineo — APPROVED** [review]

**Review by TienLuu-Saritasa — APPROVED** [review]

File: apps/react/src/api/index.ts [hunk 10]
```diff
@@ -1,7 +1,14 @@
...
 
-export const http = axios.create({
+/**
+ * HTTP client instance for making API requests.
+ *
+ * This Axios instance is pre-configured with the base URL from the application configuration.
+ *
+ * @type {AxiosInstance}
+ */
+export const http: AxiosInstance = axios.create({
```
Comment (artem-sotnichenko-saritasa) [inline apps/react/src/api/index.ts:12]: I don't think you need to write such detailed jsdoc for everything. Basically, you should explain only difficult, complex, or hacky code. 
And also, types are inherited from typescript, so there's no point to duplicate it in jsdoc

File: apps/react/src/features/genres/components/GenreCard/GenreCard.tsx [hunk 11]
```diff
@@ -3,18 +3,26 @@ import { Genre } from '@js-camp/core/models/genre';
...
+type Props = {
 
 	/** Genre. */
 	readonly genre: Genre;
-}
+};
 
 /** Card with genre data. */
-const GenreCardComponent: FC<Props> = ({ genre }) => (
+const GenreCardComponent: FC<Props> = ({ genre }: Props) => (
```
Comment (artem-sotnichenko-saritasa) [inline apps/react/src/features/genres/components/GenreCard/GenreCard.tsx:13]: `FC<Props>` is enough.

```suggestion
const GenreCardComponent: FC<Props> = ({ genre }) => (
```

**Review by artem-sotnichenko-saritasa — APPROVED** [review]

File: apps/react/src/features/genres/components/GenreCard/GenreCard.tsx (hunk 11 above)
Comment (manpham-saritasa) [reply apps/react/src/features/genres/components/GenreCard/GenreCard.tsx:13]: Apparently, if I only use `FC<Props>`, and run the script `npm run react:lint`, I will have this error message: 

> Cannot read properties of undefined (reading 'typeAnnotation')
Occurred while linting .../camp-js-2024-man-pham/apps/react/src/features/genres/components/GenreCard/GenreCard.tsx:12
Rule: "jsdoc/require-param"
Pass --verbose to see the stacktrace.

So my VN team decided to put `{ genre }: Props` to prevent this error

File: apps/react/src/api/index.ts (hunk 10 above)
Comment (manpham-saritasa) [reply apps/react/src/api/index.ts]: Thank you! The problem has been resolved
